File: DESC_Python/src/sacstuff/MatSAC/fget_sac_cut.py
import numpy as np

from .fget_sac import fget_sac


def fget_sac_cut(filename, tstart=None, tend=None):
    """
    Read SAC and cut the data by time window [tstart, tend].

    Returns (t_cut, data_cut, SAChdr) and updates SAChdr['times']['npts'].
    """
    t, data, SAChdr = fget_sac(filename)
    if tstart is None:
        tstart = -1e6
    if tend is None:
        tend = 1e6

    tmin = float(np.min(t))
    tmax = float(np.max(t))
    if tstart < tmin:
        tstart = tmin
    if tend > tmax:
        tend = tmax

    mask = (t >= tstart) & (t <= tend)
    t_cut = t[mask]
    data_cut = data[mask]

    # Update npts in header struct
    if 'times' in SAChdr:
        SAChdr['times']['npts'] = int(len(data_cut))

    return t_cut, data_cut, SAChdr

File: DESC_MATLAB/src/sacstuff/MatSAC/fget_sac_cut.m
function [t_cut,data_cut,SAChdr] = fget_sac_cut(filename,tstart,tend)
%[t,data,SAChdr] = fget_sac(filename)

% read sac into matlab
% written by Zhigang Peng
% program called
% [head1, head2, head3, data]=sac(filename);
% [SAChdr]=sachdr(head1, head2, head3);

% Updated Mon Jul 30 11:21:24 PDT 2001
% This version differs from fget_sac in that in ask
% you to input start and end time to cut the data.

if nargin <3, tend = 1000000; end
if nargin <2, tstart = -1000000; end
if nargin <1, error('ERROR!! No input file name'); end

[head1, head2, head3, data]=sac(filename);
[SAChdr]=sachdr(head1, head2, head3);
t = [SAChdr.times.b:SAChdr.times.delta:(SAChdr.data.trcLen-1)*SAChdr.times.delta+SAChdr.times.b]';
tmin = min(t);
tmax = max(t);
if tstart<tmin, tstart=tmin; end
if tend>tmax, tstart=tmin; end

t_cut = t(t>=tstart & t<=tend);
data_cut = data(t>=tstart & t<=tend);
SAChdr.times.npts = length(data_cut);
